akka-cluster-sharding/src/main/scala/akka/cluster/sharding/external/ExternalShardAllocationStrategy.scala

```diff
@@ -0,0 +1,219 @@
+        val newLocations = c.get(key).entries
+        currentLocations ++= newLocations
+        log.debug("Received updated shard locations [{}] all locations are now [{}]", newLocations, currentLocations)
+      case GetShardLocation(shard) =>
+        log.debug("GetShardLocation [{}]", shard)
+        val shardLocation = currentLocations.get(shard).map(asStr => AddressFromURIString(asStr))
+        sender() ! GetShardLocationResponse(shardLocation)
+      case GetShardLocations =>
+        log.debug("GetShardLocations")
+        sender() ! GetShardLocationsResponse(currentLocations.transform((_, asStr) => AddressFromURIString(asStr)))
+    }
+
+  }
+}
+
+class ExternalShardAllocationStrategy(systemProvider: ClassicActorSystemProvider, typeName: String)(
+    // local only ask
+    implicit val timeout: Timeout = Timeout(5.seconds))
+    extends ShardCoordinator.StartableAllocationStrategy {
+
+  private val system = systemProvider.classicSystem
+
+  import ExternalShardAllocationStrategy._
+  import akka.pattern.ask
+  import system.dispatcher
+
+  private val log = Logging(system, classOf[ExternalShardAllocationStrategy])
+
+  private var shardState: ActorRef = _
+
+  private[akka] def createShardStateActor(): ActorRef = {
+    system
+      .asInstanceOf[ExtendedActorSystem]
+      .systemActorOf(DDataStateActor.props(typeName), s"external-allocation-state-$typeName")
+  }
+
+  private val cluster = Cluster(system)
+
+  override def start(): Unit = {
+    shardState = createShardStateActor()
+  }
+
+  override def allocateShard(
+      requester: ShardRegion,
+      shardId: ShardId,
+      currentShardAllocations: Map[ShardRegion, immutable.IndexedSeq[ShardId]]): Future[ShardRegion] = {
+
+    log.debug("allocateShard [{}] [{}] [{}]", shardId, requester, currentShardAllocations)
+
+    // current shard allocations include all current shard regions
+    (shardState ? GetShardLocation(shardId))
+      .mapTo[GetShardLocationResponse]
+      .map {
+        case GetShardLocationResponse(None) =>
+          log.debug("No specific location for shard [{}]. Allocating to requester [{}]", shardId, requester)
+          requester
+        case GetShardLocationResponse(Some(address)) =>
+          // if it is the local address, convert it so it is found in the shards
+          if (address == cluster.selfAddress) {
+            currentShardAllocations.keys.find(_.path.address.hasLocalScope) match {
+              case None =>
+                log.warning("unable to find local shard in currentShardAllocation. Using requester")
+                requester
+              case Some(localShardRegion) =>
+                log.debug("allocating to local shard")
+                localShardRegion
+            }
+          } else {
+            currentShardAllocations.keys.find(_.path.address == address) match {
+              case None =>
+                log.warning(
+                  "External shard location [{}] for shard [{}] not found in members [{}]",
```

Comment:
when can this happen? is it when a member is removed from the cluster but it hasn't been updated in the LWWMap yet? then maybe not warning level?

Reply:
Yes I think so. This could happen for a short time each time a node leaves the cluster as entries aren't deleted, just replaced to where ever the shard is moved